File: dec_system/gesture_execution/gesture_execution/gesture_execution_application.py
#!/usr/bin/env python3

""" gesture_execution_application.py

Entry point for the GestureExecutionSystem lifecycle node.
Running this node starts an action server that drives Pepper through deictic,
iconic, bowing, and nodding gestures.

This node loads gesture descriptors and topic mappings from YAML, computes the
joint trajectories needed to perform a requested gesture (using inverse
kinematics for deictic pointing), and streams the resulting joint-angle
trajectory to the robot. While a gesture is executing, it periodically reports
elapsed time back to the action client so callers can stay synchronised with
the robot's motion. Deictic gestures also publish RViz markers showing the
target point, shoulder position, and pointing vector.

Subscribers:
    /joint_states (sensor_msgs/JointState)
        Current joint positions, used to track the robot's arm/head/leg state.
    /robot_localization/pose (geometry_msgs/Pose2D)
        Current robot pose in the world frame, used to compute pointing direction.

Publishers:
    /joint_angles_trajectory (naoqi_bridge_msgs/JointAnglesTrajectory)
        Joint angle trajectories sent to the robot to perform gestures.
    /gesture_execution/visualization (visualization_msgs/Marker)
        Markers visualizing deictic gesture targets, shoulder, and pointing arrow.

Actions:
    /gesture_execution (dec_interfaces/action/Gesture)
        Executes a named or typed gesture (deictic, iconic, bow, nod) with
        feedback on elapsed time and a success/failure result.

Parameters (loaded from gesture_execution_configuration.yaml):
    gestureDescriptors (string, default: "gesture.yaml")
    robotTopics (string, default: "pepperTopics.yaml")
    verboseMode (bool, default: true)

Author: Yohannes Tadesse Haile
Affiliation: Carnegie Mellon University Africa
Email: yohatad123@gmail.com
Date: April 18, 2025
Version: v1.0
"""

import rclpy
from .gesture_execution_implementation import GestureExecutionSystem

SOFTWARE_VERSION = "v1.0"

def main(args=None):
    """
    Main function to run the gesture execution system.
    """
    rclpy.init(args=args)
    
    node_name = "gesture_execution"
    
    # Construct the copyright message
    copyright_message = (
        f"{node_name} {SOFTWARE_VERSION}\n"
        "\t\t\t    This program comes with ABSOLUTELY NO WARRANTY."
    )
    
    print(copyright_message)

    gesture_system = None  # Initialize to None
    
    try:
        gesture_system = GestureExecutionSystem()
        
        # Spin to handle service calls
        rclpy.spin(gesture_system)
        
    except KeyboardInterrupt:
        print("\nShutting down gesture execution system...")
    except Exception as e:
        print(f"Error running gesture execution system: {e}")
        import traceback
        traceback.print_exc()
    finally:
        # Clean shutdown
        if gesture_system is not None:
            gesture_system.destroy_node()
        
        if rclpy.ok():  # Only shutdown if not already shut down
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: gesture_execution/src/gesture_execution_application.cpp
/* gesture_execution_application.cpp
 *
 * Entry point for the GestureExecutionNode lifecycle node. Spins the
 * node single-threaded; the class itself is implemented in
 * gesture_execution_implementation.cpp.
 *
 * Subscribers:
 *   /joint_states (sensor_msgs/msg/JointState)
 *     Current joint positions, used to track the robot's arm/head/leg state.
 *   /localization/pose (nav_msgs/msg/Odometry)
 *     Absolute map->base_footprint robot pose from fast_lio's
 *     fastlio_localization, used to compute pointing direction. The topic
 *     name comes from the RobotPose key in data/pepper_topics.yaml.
 *
 * Publishers:
 *   /joint_angles_trajectory (naoqi_bridge_msgs/msg/JointAnglesTrajectory)
 *     Joint angle trajectories sent to the robot to perform gestures.
 *   /gesture_execution/visualization (visualization_msgs/msg/Marker)
 *     Markers visualizing deictic gesture targets, shoulder, and pointing
 *     arrow.
 *
 * Actions:
 *   /gesture_execution (dec_interfaces/action/Gesture)
 *     Server that executes a named or typed gesture (deictic, iconic, bow,
 *     nod) with feedback on elapsed time and a success/failure result.
 *
 * Parameters (config/gesture_execution_configuration.yaml, under
 * gesture_action_server/ros__parameters):
 *   verbose_mode (bool, default: false)
 *   Gesture and topic data always load from fixed paths (data/gesture.yaml,
 *   data/pepper_topics.yaml) — not configurable via parameters.
 *
 * Lifecycle:
 *   configure  -> read parameters, load the gesture and topic YAML data, and
 *                 create the trajectory/marker publishers and action server
 *   activate   -> activate the publishers and subscribe to /joint_states and
 *                 /localization/pose
 *   deactivate -> destroy the joint-state and pose subscriptions and
 *                 deactivate the publishers
 *   cleanup    -> destroy the lifecycle publishers and the action server
 *   shutdown   -> log that the node is shutting down
 *
 * Author: Yohannes Tadesse Haile
 * Affiliation: Carnegie Mellon University Africa
 * Email: yohatad123@gmail.com
 * Date: July 5, 2026
 * Version: v1.0
 *
 * Copyright (C) 2025 Carnegie Mellon University Africa
 * This software is provided 'as-is' for research and educational purposes
 * within the DEC project.
 */

#include "gesture_execution/gesture_execution_interface.h"

#include "dec_common/node_runner.h"

int main(int argc, char** argv) {
    return dec_common::runNode<GestureExecutionNode>(
        argc, argv,
        {"gesture_execution v1.0 — This program comes with ABSOLUTELY NO WARRANTY.", "gesture_execution"});
}
